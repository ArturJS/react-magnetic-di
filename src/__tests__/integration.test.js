/* eslint-env jest */

import React, { Component, Fragment } from 'react';
import { mount } from 'enzyme';
import { provideDependencies, settings, DependencyProvider } from '../index';

jest.mock('../settings');

class Label extends Component {
  static dependencies = provideDependencies({
    Text: () => '',
    Wrapper: ({ children }) => children,
  });
  render() {
    const { Wrapper, Text } = Label.dependencies();
    return (
      <Wrapper>
        <Text />
      </Wrapper>
    );
  }
}

class Input extends Component {
  static dependencies = provideDependencies({
    Text: () => '',
  });
  render() {
    const { Text } = Input.dependencies();
    return <Text />;
  }
}

const TextMock = () => '';
const TextMockAlt = () => '';
const WrapperMock = ({ children }) => children;

describe('Integration', () => {
  it('should return real dependencies if provider less', () => {
    settings.enabled = false;
    const wrapper = mount(
      <Fragment>
        <Label />
        <Input />
      </Fragment>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should return real dependencies if disabled', () => {
    settings.enabled = false;
    const wrapper = mount(
      <DependencyProvider use={{ Text: TextMock }}>
        <Label />
        <Input />
      </DependencyProvider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should override all dependencies with same name', () => {
    settings.enabled = true;
    const wrapper = mount(
      <DependencyProvider use={{ Text: TextMock }}>
        <Label />
        <Input />
      </DependencyProvider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should override per type dependencies', () => {
    settings.enabled = true;
    const wrapper = mount(
      <DependencyProvider target={Label} use={{ Text: TextMock }}>
        <Label />
        <Input />
      </DependencyProvider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should allow override composition', () => {
    settings.enabled = true;
    const wrapper = mount(
      <DependencyProvider use={{ Text: TextMock, Wrapper: WrapperMock }}>
        <DependencyProvider target={Label} use={{ Text: TextMockAlt }}>
          <Label />
          <Input />
        </DependencyProvider>
      </DependencyProvider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });
});
