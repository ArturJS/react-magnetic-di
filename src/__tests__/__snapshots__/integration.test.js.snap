// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration should allow override composition 1`] = `
"<DependencyProvider use={{...}}>
  <DependencyProvider target={[Function: Label]} use={{...}}>
    <Label>
      <WrapperMock>
        <TextMockAlt />
      </WrapperMock>
    </Label>
    <Input>
      <TextMock />
    </Input>
  </DependencyProvider>
</DependencyProvider>"
`;

exports[`Integration should override all dependencies with same name 1`] = `
"<DependencyProvider use={{...}}>
  <Label>
    <Wrapper>
      <TextMock />
    </Wrapper>
  </Label>
  <Input>
    <TextMock />
  </Input>
</DependencyProvider>"
`;

exports[`Integration should override per type dependencies 1`] = `
"<DependencyProvider target={[Function: Label]} use={{...}}>
  <Label>
    <Wrapper>
      <TextMock />
    </Wrapper>
  </Label>
  <Input>
    <Text />
  </Input>
</DependencyProvider>"
`;

exports[`Integration should return real dependencies if disabled 1`] = `
"<DependencyProvider use={{...}}>
  <Label>
    <Wrapper>
      <Text />
    </Wrapper>
  </Label>
  <Input>
    <Text />
  </Input>
</DependencyProvider>"
`;

exports[`Integration should return real dependencies if provider less 1`] = `
"<Label>
  <Wrapper>
    <Text />
  </Wrapper>
</Label>


<Input>
  <Text />
</Input>"
`;
